REP: Coordinate Frames for Serial Industrial Manipulators (WIP) (#3)
gavanderhoorn/rep: rep-0199.rst

REP 105 [#REP105]_ and REP 120 [#REP120]_ already define frames for mobile bases and humanoid robots.
This REP extends the set of standardised frames in ROS and labels important locations on serial industrial manipulators such as the origin of the main Cartesian coordinate system, the flange and the tool centre point.
Finally, it provides naming guidelines for intermediate link frames and application-specific tool frames.

Definitions
===========

End Effector (EEF)
    A device at the end of a robotic arm, designed to interact with the environment.
    The exact nature of this device depends on the application of the robot.
Tool Centre Point (TCP)
    Point in relation to which all robot positioning is defined or the origin of the tool's coordinate frame.

Assumptions
===========

#. The key words *must*, *must not*, *required*, *shall*, *shall not*, *should*, *should not*, *recommended*, *may*, and *optional* in this document are to be interpreted as defined in [#RFC2119]_.

Specification
=============

Coordinate Frames
-----------------

base_link
'''''''''

> The coordinate frame called ``base_link`` shall be rigidly attached to the robot root body.

Reviewer: Which frames are optional and which ones mandatory? What happens, if a robot is missing some of the frames? Are there default assumtions such as `base=base_link`, `tool0=flange`? This would allow easier transitions of existing robot models.

```suggestion
The robot model shall define a  coordinate frame called ``base_link``.
```

Author: > Which frames are optional and which ones mandatory?

I'll update the REP, but only `base`, `flange` and `tool0` are mandatory. The `link_n` frame names are a result of an earlier review of the REP in which we discussed whether introducing naming conventions for intermediate frames (ie: those of the links) would make sense. In the end I decided that, yes, it does make sense -- for the reasons stated in the rationale paragraph of the `link_n` section.

> What happens, if a robot is missing some of the frames?

similar to what would happen if you intend to reuse the navigation stack with a robot that does not have an `odom` frame: things will fail, as the navstack requires a REP-105 compliant system.

> Are there default assumtions such as `base=base_link`, `tool0=flange`? This would allow easier transitions of existing robot models.

While the idea of a default is nice, I'm curious how you see this working. The REP attempts to standardise urdf/xacro `link` / TF frame names. There is no facility in either of those for defaults or fallbacks.

I would also not be very comfortable if there were: without these frames present, a TF lookup `base -> tool0` could evaluate down to `base_link -> flange` (if `flange` exists), which is an entirely different lookup and the result would be completely different, semantically, from the intended lookup (as `base_link` could be anywhere and has no orientation constraints and `flange` intentionally does not correspond to `tool0`).

Finally: as I wrote above: the only frames that are *really* required are `base` and `tool0` (and potentially `flange`). If we go with the idea of defaults, just about every robot model could start claiming compliance with this REP after adding a `link` with the name `flange` somewhere. That is not the intention of the REP.

---

These are good questions btw, and I'll see how I can incorporate them (and their answers) into the REP proper.

  base_link
  ├ base
  ├ left_base_link
  │ ├ left_base
  │ └ left_link_1
  │   └ ..
  │     └ left_link_N
  │       └ left_flange
  │         └ left_tool0
  └ right_base_link
    ├ right_base
    └ right_link_1
      └ ..
        └ right_link_N
          └ right_flange
            └ right_tool0

Note that ``base_link`` in this example is the root of the entire robot structure and should be used when integrating the robot into a larger assembly.

Vendor Nomenclature Mapping
===========================

This section shows a mapping between vendor-specific frame nomenclature and the ``base`` and ``tool0`` frames as defined by this REP.

Note that for most vendors, ``tool0`` corresponds to an *all-zeros* tool frame configuration as described in the `tool0`_ section.
The names listed here in the *Vendor Name* column refer to the generic names for frames as used in the documentation of the control systems of the respective vendors, and not necessarily to any specific configurations of those frames.

+------------------+------------+---------------+
|                  | This REP   | Vendor Name   |
+==================+============+===============+

> | ABB              | ``base``   | Base          |

Reviewer: I'd keep this list out of the REP and put it into the wiki. The wiki page is easier to maintain for future manufacurers and changes in vendor-specific implementations. Here a link to the wiki page would suffice.

Author: >  The wiki page is easier to maintain for future manufacurers and changes in vendor-specific implementations

While I agree that a wiki would be easier to edit, the number of changes to this table will probably be very few. Industrial robot OEMs are not likely to change the coordinate frames they use in their controllers (because of legacy, documentation, conventions, training, etc), so this table is only expected to grow, not change.

However, I'm not entirely sold on having this table in the REP either, so I'll count your comment as a :-1: and wait for some more feedback.
